examples/sac/sac.py

```diff
@@ -48,132 +51,164 @@ def main(cfg: "DictConfig"):  # noqa: F821
         collected_frames += current_frames
 
-        # optimization steps
+        # Optimization steps
+        training_start = time.time()
         if collected_frames >= init_random_frames:
-            (actor_losses, q_losses, alpha_losses) = ([], [], [])
-            for _ in range(num_updates):
-                # sample from replay buffer
+            losses = TensorDict(
+                {},
+                batch_size=[
+                    num_updates,
+                ],
+            )
+            for i in range(num_updates):
+                # Sample from replay buffer
                 sampled_tensordict = replay_buffer.sample().clone()
 
+                # Compute loss
                 loss_td = loss_module(sampled_tensordict)
 
                 actor_loss = loss_td["loss_actor"]
                 q_loss = loss_td["loss_qvalue"]
                 alpha_loss = loss_td["loss_alpha"]
-                loss = actor_loss + q_loss + alpha_loss
 
-                optimizer.zero_grad()
-                loss.backward()
-                optimizer.step()
+                # Update actor
+                optimizer_actor.zero_grad()
+                actor_loss.backward()
+                optimizer_actor.step()
+
+                # Update critic
+                optimizer_critic.zero_grad()
+                q_loss.backward()
+                optimizer_critic.step()
+
+                # Update alpha
+                optimizer_alpha.zero_grad()
+                alpha_loss.backward()
+                optimizer_alpha.step()
 
-                q_losses.append(q_loss.item())
-                actor_losses.append(actor_loss.item())
-                alpha_losses.append(alpha_loss.item())
+                losses[i] = loss_td.select(
+                    "loss_actor", "loss_qvalue", "loss_alpha"
+                ).detach()
 
-                # update qnet_target params
+                # Update qnet_target params
                 target_net_updater.step()
 
-                # update priority
+                # Update priority
                 if prb:
                     replay_buffer.update_priority(sampled_tensordict)
 
-        rewards.append(
-            (i, tensordict["next", "reward"].sum().item() / env_per_collector)
-        )
-        train_log = {
-            "train_reward": rewards[-1][1],
-            "collected_frames": collected_frames,
-        }
-        if q_loss is not None:
-            train_log.update(
-                {
-                    "actor_loss": np.mean(actor_losses),
-                    "q_loss": np.mean(q_losses),
-                    "alpha_loss": np.mean(alpha_losses),
-                    "alpha": loss_td["alpha"],
-                    "entropy": loss_td["entropy"],
-                }
+        training_time = time.time() - training_start
+        episode_rewards = tensordict["next", "episode_reward"][
+            tensordict["next", "done"]
+        ]
+
+        # Logging
+        if len(episode_rewards) > 0:
```

Comment:
Shouldn't we group all the logging in a single function, to avoid overloading the training loop?

Reply:
I agree, if you can have a look at the DDPG PR I tried to compress the logging but I'm open to other ideas to do it.